tests/qos/test_buffer.py

```diff
@@ -0,0 +1,848 @@
+
+
+@pytest.fixture(scope="module", autouse=True)
+def setup_module(duthost):
+    detect_buffer_model(duthost)
+    if buffer_model_dynamic:
+        detect_ingress_pool_number(duthost)
+        load_lossless_headroom_data(duthost)
+        load_test_parameters(duthost)
+
+        logging.info("Got cable length: default {}".format(default_cable_length_list))
+        logging.info("Got ingress pool number {}".format(default_ingress_pool_number))
+        logging.info("Got lossless headroom data {}".format(default_lossless_headroom_data))
+    else:
+        pytest.skip("Dynamic buffer isn't enabled, skip the test")
+
+    yield
+
+
+def check_pool_size(duthost, ingress_lossless_pool_oid, **kwargs):
+    if duthost.facts['asic_type'] == 'mellanox':
+        old_headroom = int(kwargs["old_headroom"])
+
+        if "old_pg_number" in kwargs:
+            old_pg_number = int(kwargs["old_pg_number"])
+        else:
+            old_pg_number = 2
+
+        if "new_pg_number" in kwargs:
+            new_pg_number = int(kwargs["new_pg_number"])
+        else:
+            new_pg_number = old_pg_number
+
+        if new_pg_number:
+            if "new_headroom" in kwargs:
+                new_headroom = int(kwargs["new_headroom"])
+            else:
+                new_headroom = old_headroom
+            new_reserved = new_pg_number * new_headroom
+        else:
+            new_reserved = 0
+
+        curr_pool_size = int(kwargs["pool_size"])
+
+        original_memory = curr_pool_size * default_ingress_pool_number + old_headroom * old_pg_number
+        expected_pool_size = (original_memory - new_reserved) / default_ingress_pool_number
+        pool_size = duthost.shell('redis-cli hget "BUFFER_POOL_TABLE:ingress_lossless_pool" size')['stdout']
+        pytest_assert(int(pool_size) == expected_pool_size, "Pool size isn't correct: expected {} but got {}".format(expected_pool_size, pool_size))
+
+        if ingress_lossless_pool_oid:
+            pool_sai = _compose_dict_from_cli(duthost.shell('redis-cli -n 1 hgetall ASIC_STATE:SAI_OBJECT_TYPE_BUFFER_POOL:{}'.format(ingress_lossless_pool_oid))['stdout'].split('\n'))
+            pytest_assert(int(pool_sai['SAI_BUFFER_POOL_ATTR_SIZE']) == expected_pool_size)
+
+
+def check_pg_profile(duthost, pg, expected_profile):
+    profile = duthost.shell('redis-cli hget {} profile'.format(pg))['stdout'][1:-1]
+    pytest_assert(profile == 'BUFFER_PROFILE_TABLE:' + expected_profile, 'Expected profile {} not found'.format(expected_profile))
+
+
+def check_pfc_enable(duthost, port, expected_pfc_enable_map):
+    pfc_enable = duthost.shell('redis-cli -n 4 hget "PORT_QOS_MAP|{}" pfc_enable'.format(port))['stdout']
+    pytest_assert(expected_pfc_enable_map == pfc_enable, \
+        "Expected pfc enable map {} doesn't match {}".format(expected_pfc_enable_map, pfc_enable))
+
+
+def check_lossless_profile_removed(duthost, profile, sai_oid = None):
+    time.sleep(10)
```

Comment:
why need to sleep? if this is done by another method and you are assuming it was not completed that i believe you should validate the status and only then sleep. you can even sleep less if this is already done.

Reply:
I think we can take advantage of "wait_until".